README
NAME
    Object::Util - a selection of utility methods that can be called on
    blessed objects

SYNOPSIS
       use Object::Util;
   
       # $foo might be undef, but this should not die
       if ($foo->$_isa("Bar")) {
          ...;
       }

DESCRIPTION
    This module is inspired by Safe::Isa, Object::Tap, and my own OO
    experiences. It is a hopefully helpful set of methods for working with
    objects, exposed as lexical coderef variables.

  Rationale
    Providing methods as coderefs so that you can do:

       $object->$_foo(@args)

    ... is unusual, so probably requires some explanation.

    Firstly some of these methods are designed to be called on either a
    blessed object or some kind of unblessed reference or value. Calling a
    method on an unblessed reference like this will croak:

       $ref->foo(@args)

    Ditto calling methods on undef. Coderefs don't suffer from that problem.

    More importantly though, the aim of this module is that these methods
    should be available for you to call on *any* object. You can only call
    `$object->foo(@args)` if $object's class implements a method called `foo`,
    or inherits from a superclass that does. Coderef methods can be called on
    any object.

    This module adopts the $_foo naming convention pioneered by modules such
    as Safe::Isa. However (unlike Safe::Isa) the coderefs it provides are
    *true lexical variables* (a.k.a. `my` variables), not package variables
    (a.k.a. `our` variables).

  Methods
   Object construction and manipulation
    $_new
        Can be used like `$class->$_new(@args)` to create a new object.
        Object::Util will use Module::Runtime to *attempt* to load $class if
        it is not already loaded. $class is expected to provide a method
        called `new`.

        Can also be used as `$factory->$_new(@args)` to create a new object,
        where $factory is a coderef or an object overloading `&{}`. In this
        case, $_new will simply call `$factory->(@args)` and expect that to
        return an object.

    $_clone
        If the object provides a `clone` method, calls that. Or if the object
        appears to be Moose- or Mouse-based, clones it using the metaobject
        protocol.

        Otherwise takes the naive approach of treating the object as a hashref
        of attribute values, and creates a new object of the same class.

           # clone overrides some attributes from the original object
           my $glenda = $glen->$_clone(name => "Glenda", gender => "f");

        That final fallback obviously massively breaks your class'
        encapsulation, so it should be used sparingly.

    $_with_traits
        Calling `$class->$_with_traits(@traits)` will return a new class name
        that does some extra traits. Should roughly support Moose, Mouse, Moo,
        and Role::Tiny, though combinations of frameworks (e.g. consuming a
        Moose role in a Mouse class) will not always work.

        If $class is actually a (factory) coderef, then this will only
        *partly* work. Example:

           my $factory  = sub { Foo->new(@_) };
           my $instance = $factory->$_with_traits("Bar")->$_new(%args);

        The object $instance should now be a `Foo` object, and should do the
        `Bar` role, however if `Bar` defines any *attributes*, then $_new will
        not have initialized them correctly. This is because of the opacity of
        the $factory: $_with_traits cannot peek inside it and apply traits to
        the `Foo` class; instead it needs to build $instance and apply the
        traits to the already-built object. Therefore any behaviour that `Bar`
        defines for the constructor will have been ignored.

        It is sometimes possible to work around this issue using:

           my $factory  = sub { Foo->new(@_) };
           my $instance = $factory->$_with_traits("Bar")->$_new(%args);
           $instance = $instance->$_clone(%args);

    $_extend
        Calling `$object->$_extend(\@traits, \%methods)` will add some extra
        roles and/or methods to an existing object.

        Either @traits or %methods may be omitted. For example,
        `$object->$_extends(\@traits)` or `$object->$_extends(\%methods)`.
        `$object->$_extends()` also works fine, and is a no-op.

        Like Object::Extend, but with added support for roles.

   Method call modifiers
    $_call_if_object
        `$object->$_call_if_object($method => @args)` works like
        `$object->$method(@args)`, but if $object is undefined, returns
        `undef` instead of throwing an exception.

        $method may be a method name, or a coderef (anonymous method).

        Same as Safe::Isa.

    $_try
        `$object->$_try($method => @args)` works like
        `$object->$method(@args)`, but if *any exception is thrown* returns
        `undef` instead.

        $method may be a method name, or a coderef (anonymous method).

    $_tap
        `$object->$_tap($method => @args)` works like
        `$object->$method(@args)`, but discards the method's return value
        (indeed it calls the method in void context), and instead returns the
        object itself, making it useful for chaining.

        $method may be a method name, or a coderef (anonymous method).

        Same as Object::Tap, or the `tap` method in Ruby.

    $_isa
        `$object->$_isa($class)` works like `isa` as defined in UNIVERSAL, but
        if $object is undefined, returns false instead of throwing an
        exception.

        A shortcut for `$object->$_call_if_method(isa => $class)`.

        Same as Safe::Isa.

    $_does
        `$object->$_does($role)` works like `does` as defined in
        Moose::Object, but if $object is undefined, returns false instead of
        throwing an exception.

        A shortcut for `$object->$_call_if_method(does => $role)`.

        Same as Safe::Isa.

    $_DOES
        `$object->$_DOES($role)` works like `DOES` as defined in UNIVERSAL,
        but if $object is undefined, returns false instead of throwing an
        exception.

        A shortcut for `$object->$_call_if_method(DOES => $role)`.

        Same as Safe::Isa.

    $_can
        `$object->$_can($method)` works like `can` as defined in UNIVERSAL,
        but if $object is undefined, returns `undef` instead of throwing an
        exception.

        There is one other significant deviation from the behaviour of
        UNIVERSAL's `can` method: $_can also returns true if $method is an
        unblessed coderef. (The behaviour of $method if it is a blessed object
        -- particularly in the face of overloading -- can be unintuitive, so
        is not supported by $_can.)

        Similar to Safe::Isa, but not quite the same.

   Object utility methods
    $_dump
        Calling `$object->$_dump` returns a Data::Dumper dump of the object,
        with some useful changes to the default Data::Dumper output. (Same as
        Data::Dumper::Concise.)

        If the object provides its own `dump` method, this will be called
        instead.

  Implementation Details
    B::Hooks::Parser is used to inject these methods into your lexical scope,
    and `Internals::SvREADONLY` (an internal function built into the Perl
    core) is used to make them read-only, so you can't do:

       use Object::Util;
       $_isa = sub { "something else" };

    If this module detects that B::Hooks::Parser cannot be used on your
    version of Perl, or your Perl is too old to have Internals::SvREADONLY,
    then it has various fallback routes, but the variables it provides may end
    up as package (`our`) variables, or not be read-only.

    If the magic works on your version of Perl, but you wish to avoid the
    magic anyway, you can switch it off:

       use Object::Util magic => 0;

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=Object-Util>.

SEE ALSO
    Safe::Isa, UNIVERSAL, Object::Tap, MooseX::Clone, Data::Dumper::Concise,
    Object::Extend.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2014 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Object-Util
===========

Created:      2014-09-08
Home page:    <https://metacpan.org/release/Object-Util>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Object-Util>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.005	2014-09-16

 [ Documentation ]
 - Restructure and further clarify documentation of methods.

 [ Other ]
 - Clean up some old `use` statements that were no longer needed.
 - Delay loading Module::Runtime until it's actually needed.
 - Make it possible to avoid the B::Hooks::Parser/Internals::SvREADONLY
   magic.

0.004	2014-09-16

 [ Documentation ]
 - Document that most places $method is used in the documentation accept
   either a method name or a coderef.
 - Fix referenes to $object->$_does($class) (and $_DOES too) where $class
   should have been $role.

 [ Other ]
 - Support $object->$_can($method) where $method is a coderef.

0.003	2014-09-16

 - Allow $_with_traits to (mostly) work with factory coderefs.
 - If an object provides its own dump method, then $_dump should defer to
   that.

0.002	2014-09-09

 [ Test Suite ]
 - Fix Mouse-related failures.

 [ Other ]
 - Added: Add $_extends method.

0.001	2014-09-09	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "a selection of utility methods that can be called on blessed objects",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 1,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.021, CPAN::Meta::Converter version 2.140640",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "Object-Util",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { suggests => { "Dist::Inkt" => 0.001 } },
                        runtime   => {
                                       requires => {
                                         "List::Util"         => 1.29,
                                         "Module::Runtime"    => 0,
                                         "MooX::Traits::Util" => 0,
                                         "perl"               => "5.006000",
                                         "Role::Tiny"         => 0,
                                         "Scalar::Util"       => 1.23,
                                       },
                                     },
                        test      => {
                                       recommends => { Moo => "1.000000", Moose => "2.0000" },
                                       requires   => {
                                                       "Test::Fatal"           => 0,
                                                       "Test::More"            => 0.96,
                                                       "Test::Requires"        => 0,
                                                       "Test::Warnings"        => 0,
                                                       "Test::Without::Module" => 0,
                                                     },
                                       suggests   => { Mouse => "1.00" },
                                     },
                      },
  "provides"       => {
                        "Object::Util" => { file => "lib/Object/Util.pm", version => 0.005 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => { web => "http://rt.cpan.org/Dist/Display.html?Queue=Object-Util" },
                        homepage     => "https://metacpan.org/release/Object-Util",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          url  => "git://github.com/tobyink/p5-object-util.git",
                                          web  => "https://github.com/tobyink/p5-object-util",
                                        },
                        x_identifier => "http://purl.org/NET/cpan-uri/dist/Object-Util/project",
                      },
  "version"        => 0.005,
};

my %dynamic_config;
do {
$meta->{prereqs}{runtime}{requires}{'B::Hooks::Parser'} = '0.10' if $] >= 5.008001;
$meta->{prereqs}{runtime}{requires}{'UNIVERSAL::DOES'}  = 0      if $]  < 5.010000;

};

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

